ttg/ttg/wrap.h

```diff
@@ -18,7 +18,10 @@ class WrapOp
   using input_edges_type = typename baseT::input_edges_type;
   using output_edges_type = typename baseT::output_edges_type;
 
-  std::function<boost::callable_traits::function_type_t<funcT>> func;
+  using noref_funcT = typename std::remove_reference_t<funcT>;
+  typename std::conditional<std::is_function<noref_funcT>::value,
```

Comment:
now makes sense why I used `std::function`: it is an owning handle, and copies callable when given a reference to it (passing `std::reference_wrapper` to a callable is the proper way to have to store a reference and not a copy).

In my experience sometimes it is desirable to have the user manage lifetime of the callable rather than have the task engine do it ... e.g. when callables are not copyable/movable (e.g. singletons, or for other reasons). So it does not make sense to always copy. Perhaps it makes sense to do this:
- `WrapOp<funcT>` is not valid with `funcT` a reference type (static_assert it)
- `wrap` handles the logic of deciding how to hold the callable. It could even wrap the callable in `std::function<>`, but probably makes sense to use `std::function_ref` for callables passed by reference (I have its implementation copied from somewhere [here](https://github.com/ValeevGroup/tiledarray/blob/master/src/TiledArray/util/function.h#L48) and there are [others](https://github.com/TartanLlama/function_ref)) and just the value type for the callables that are passed by rvalue ref.

Reply:
Coming back to this after a while... I might be misreading your comment but I think we always need to take a copy of the callable because we cannot make any assumptions about the life-time of the callable object. This code for example will go up in flames if we don't take a copy:

```C
auto make_x(...) {
  auto fn = [](){ /* ... */ };
  return ttg::make_tt(fn, ...);
}
```

If we simply store a reference to `fn`, it will be out of scope and destroyed before we call it and we end up with a dangling reference. We should leave it to the user to pass us a `reference_wrapper` if his callable is not copyable.

Reply:
OK, having the user pass `reference_wrapper` for shallow-copy semantics is reasonable.